benchmarks/nacl.cc

```diff
@@ -0,0 +1,192 @@
+#include "util.h"
+
+#include "Hacl_NaCl.h"
+
+#define HACL_NACL_SECRETKEYBYTES 32
+#define HACL_NACL_PUBLICKEYBYTES 32
+#define HACL_NACL_BEFORENMBYTES 32
+#define HACL_NACL_KEYBYTES 32
+#define HACL_NACL_NONCEBYTES 24
+#define HACL_NACL_MACBYTES 16
+
+using namespace std;
+
+static bytes alice_pk(HACL_NACL_PUBLICKEYBYTES, 0);
+static bytes alice_sk(HACL_NACL_SECRETKEYBYTES, 0);
+static bytes bob_pk(HACL_NACL_PUBLICKEYBYTES, 0);
+static bytes bob_sk(HACL_NACL_SECRETKEYBYTES, 0);
+static bytes plaintext(1000, 0x37);
+static bytes ciphertext(HACL_NACL_MACBYTES + plaintext.size());
+static bytes decrypted(plaintext.size(), 0);
+static bytes nonce(HACL_NACL_NONCEBYTES, 0);
+static bytes tag(HACL_NACL_MACBYTES, 0);
+static bytes k(HACL_NACL_BEFORENMBYTES, 0);
+static bytes key(HACL_NACL_KEYBYTES, 0);
+
+static void
+Hacl_NaCl_oneshot_combined(benchmark::State& state)
+{
+  crypto_box_keypair_alice(alice_sk.data(), alice_pk.data());
+  crypto_box_keypair_bob(bob_sk.data(), bob_pk.data());
+
+  while (state.KeepRunning()) {
+    generate_random(nonce.data(), nonce.size());
```

Comment:
Question: I'm not sure where to sample the random nonce. It seems to me that using random nonces is less artificial. However, now we add the sampling time, i.e., `EverCrypt_DRBG_{create, instantiate, generate, uninstantiate}` as used in `generate_random`, to the benchmark.

My feeling is that we should move that before the loop. @franziskuskiefer, what do you think?

Reply:
Had the same thought and went with the less artificial solution. I'll wait for @franziskuskiefer

Reply:
FWIW: Here are some guidelines how to write benchmarks: https://github.com/dotnet/performance/blob/main/docs/microbenchmark-design-guidelines.md

I was worried about caching side-effects when using static data. But, according to the above guidelines, "input data [...] should be always exactly the same data".

Reply:
I moved the random nonce generation out of the loop.